lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -991,38 +863,58 @@ createAsyncOps(scf::ForOp &forOp, tt::CoarseSchedule &schedule,
   }
 
+  // FIXME: loads can be in different (stage, cluster)
   // Create two counters for the insert and extract indices to avoid creating
   // long liverange.
   builder.setInsertionPoint(newForOp.getBody(), newForOp.getBody()->begin());
+  auto [insertStage, insertCluster] = getInsertIdxStageCluster(loadToInfo);
+  auto [extractStage, extractCluster] = getExtractIdxStageCluster(loadToInfo);
   insertIdx = builder.create<arith::AddIOp>(loc, insertIdx, one);
+  setStageCluster(forOp, insertIdx.getDefiningOp(), insertStage, insertCluster);
   Value cndIns = builder.create<arith::CmpIOp>(loc, arith::CmpIPredicate::slt,
                                                insertIdx, numBuffersVal);
+  setStageCluster(forOp, cndIns.getDefiningOp(), insertStage, insertCluster);
   insertIdx = builder.create<arith::SelectOp>(loc, cndIns, insertIdx, zero);
+  setStageCluster(forOp, insertIdx.getDefiningOp(), insertStage, insertCluster);
 
+  SmallVector<Operation *> extractOps;
   extractIdx = builder.create<arith::AddIOp>(loc, extractIdx, one);
+  extractOps.push_back(extractIdx.getDefiningOp());
+  setStageCluster(forOp, extractIdx.getDefiningOp(), extractStage,
+                  extractCluster);
   Value cndExt = builder.create<arith::CmpIOp>(loc, arith::CmpIPredicate::slt,
                                                extractIdx, numBuffersVal);
+  extractOps.push_back(cndExt.getDefiningOp());
+  setStageCluster(forOp, cndExt.getDefiningOp(), extractStage, extractCluster);
   extractIdx = builder.create<arith::SelectOp>(loc, cndExt, extractIdx, zero);
+  extractOps.push_back(extractIdx.getDefiningOp());
+  setStageCluster(forOp, extractIdx.getDefiningOp(), extractStage,
+                  extractCluster);
   if (phase) {
     Value nextPhase = builder.create<arith::XOrIOp>(loc, phase, one);
+    setStageCluster(forOp, nextPhase.getDefiningOp(), extractStage,
+                    extractCluster);
     phase = builder.create<arith::SelectOp>(loc, cndExt, phase, nextPhase);
+    setStageCluster(forOp, phase.getDefiningOp(), extractStage, extractCluster);
   }
   createTMABarrierAndWait(forOp, asyncLoads, insertIdx, extractIdx, phase,
-                          numBuffers, schedule, barriers, loadToInfo);
-
-  // Create a cluster for the prefetches. It may end up being empty, but this
-  // is OK.
-  tt::CoarseSchedule::Cluster prefetchCluster = schedule.clusters.newAtBack();
+                          numBuffers, barriers, loadToInfo);
 
+  auto [_, maxClusterId] = getMinMaxCluster(forOp);
   for (AsyncLoad &asyncLoad : asyncLoads) {
     if (auto loadOp = dyn_cast<tt::LoadOp>(asyncLoad.loadOp)) {
-      createAsyncCopy(forOp, loadOp, asyncLoad.alloc, insertIdx, extractIdx,
-                      schedule, prefetchCluster, loadToInfo, numStages);
+      int retCode =
+          createAsyncCopy(forOp, loadOp, asyncLoad.alloc, insertIdx, extractIdx,
+                          loadToInfo, numStages, maxClusterId);
```

Comment:
We have set (stage, cluster) for extractIdx to the first Use of all loadOps (see helper getExtractIdxStageCluster). If wait/viewLoad are put in the prefetch stage, cluster, we need to reset the attributes for extractIdx here.
It is not clear to me if we need multiple copies of extractIdx here.